storage/badger/receipts.go

```diff
@@ -142,8 +143,78 @@ func (r *ExecutionReceipts) Index(blockID, receiptID flow.Identifier) error {
 	})
 }
 
+func (r *ExecutionReceipts) IndexByBlockIDAndExecutionID(blockID, executorID, receiptID flow.Identifier) error {
+	return operation.RetryOnConflict(r.db.Update, func(tx *badger.Txn) error {
+
+		err := operation.IndexExecutionReceiptByBlockIDExecutionID(blockID, executorID, receiptID)(tx)
+		if err == nil {
+			return nil
+		}
+
+		if !errors.Is(err, storage.ErrAlreadyExists) {
+			return err
+		}
+
+		// when trying to index a receipt for a block, and there is already a receipt indexed for this block,
+		// double check if the indexed receipt has the same result.
+		// if the result is the same, we could skip indexing the receipt
+		// if the result is different, then return error
+		var storedReceiptID flow.Identifier
+		err = operation.LookupExecutionReceiptByBlockIDExecutionID(blockID, executorID, &storedReceiptID)(tx)
+		if err != nil {
+			return fmt.Errorf("there is a receipt stored already, but cannot retrieve the ID of it: %w", err)
+		}
+
+		storedReceipt, err := r.byID(storedReceiptID)(tx)
+		if err != nil {
+			return fmt.Errorf("there is a receipt stored already, but cannot retrieve of it: %w", err)
+		}
+
+		storingReceipt, err := r.byID(receiptID)(tx)
+		if err != nil {
+			return fmt.Errorf("attempting to index a receipt, but the receipt is not stored yet: %w", err)
+		}
+
+		storedResultID := storedReceipt.ExecutionResult.ID()
+		storingResultID := storingReceipt.ExecutionResult.ID()
+		if storedResultID != storingResultID {
+			return fmt.Errorf(
+				"storing receipt that is different from the already stored one for block: %v, execution ID: %v, storing receipt: %v, stored receipt: %v. storing result: %v, stored result: %v, %w",
+				blockID, executorID, receiptID, storedReceiptID, storingResultID, storedResultID, storage.ErrDataMismatch)
+		}
+
+		return nil
+	})
+}
+
 func (r *ExecutionReceipts) ByBlockID(blockID flow.Identifier) (*flow.ExecutionReceipt, error) {
 	tx := r.db.NewTransaction(false)
 	defer tx.Discard()
 	return r.byBlockID(blockID)(tx)
 }
+
+func (r *ExecutionReceipts) ByBlockIDAllExecutionID(blockID flow.Identifier) ([]flow.ExecutionReceipt, error) {
```

Comment:
Naming was a bit hard to understand, maybe change it to AllExecutionReceipts rather than AllExecutionID?